File: duplicates/go/matrix.go

package main

import (
    "fmt"
    "flag"
    "strconv"
    "rand"
    "time"
)


type Matrix [][]float32


func MatrixMin(m [][]float32) (int, int) {
    min_i := 0
    min_j := 0
    min_v := float32(1.0)
    
    for i, r := range m {
        for j, v := range r {
            if v < min_v {
                min_i = i
                min_j = j
                min_v = v
            } 
        }
    }
    
    return min_i, min_j
}


func RandMatrix(size int) Matrix {
    m := make([][]float32, size)
    
    for i := range m {
        m[i] = make([]float32, size)
        for j := range m[i] {
            m[i][j] = rand.Float32()
        }
    }
    
    return Matrix(m)
}


func main() {
    flag.Parse()
    n, _ := strconv.Atoi(flag.Arg(0))
    
    start := time.Nanoseconds()
    m := RandMatrix(n)
    end := time.Nanoseconds()
    fmt.Printf("Creating %v x %v matrix took %vms\n", n, n, (end - start) / 1000000.0)
    
    start = time.Nanoseconds()
    i, j := MatrixMin(m)
    end = time.Nanoseconds()
    fmt.Printf("Finding min in %v x %v matrix took %vms\n", n, n, (end - start) / 1000000.0)
    fmt.Printf("Min was %f\n", m[i][j])
}

File: duplicates/go/matrix.js
function MatrixMin(m) {
    var min_i = 0;
    var min_j = 0;
    var min_v = 1.0;
    
    n = m.length;
    
    for (var i = 0; i < n; i++) {
        for (var j = 0; j < n; j++) {
            if (m[i][j] < min_v) {
                min_i = i;
                min_j = j;
                min_v = m[i][j];
            }
        }
    }

    return [min_i, min_j];
}

function RandMatrix(n) {
    m = [];
    
    for (var i = 0; i < n; i++) {
        m[i] = [];
        for (var j = 0; j < n; j++) {
            m[i][j] = Math.random();
        }
    }
    
    return m;
}

var n = parseInt(process.argv[2]);

var start = Date.now();
var m = RandMatrix(n);
var end = Date.now();
console.log("Creating " + n + " x " + n + " matrix took " + (end - start) + "ms");

start = Date.now();
var min = MatrixMin(m);
end = Date.now();
console.log("Finding " + n + " x " + n + " matrix took " + (end - start) + "ms");
console.log("Min was " + m[min[0]][min[1]]);

